test/Microsoft.SqlTools.ServiceLayer.UnitTests/QueryExecution/DataStorage/SaveAsMarkdownFileStreamWriterTests.cs

```diff
@@ -0,0 +1,383 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.SqlTools.ServiceLayer.QueryExecution.Contracts;
+using Microsoft.SqlTools.ServiceLayer.QueryExecution.DataStorage;
+using Microsoft.SqlTools.ServiceLayer.UnitTests.Utility;
+using NUnit.Framework;
+
+namespace Microsoft.SqlTools.ServiceLayer.UnitTests.QueryExecution.DataStorage
+{
+    public class SaveAsMarkdownFileStreamWriterTests
+    {
+        // Regex: Matches '|' not preceded by a '\'
+        private static readonly Regex UnescapedPipe = new Regex(@"(?<!\\)\|", RegexOptions.Compiled);
+
+        [Test]
+        public void Constructor_NullStream()
+        {
+            // Act
+            TestDelegate action = () => _ = new SaveAsMarkdownFileStreamWriter(
+                null,
+                new SaveResultsAsMarkdownRequestParams(),
+                Array.Empty<DbColumnWrapper>()
+            );
+
+            // Assert
+            Assert.Throws<ArgumentNullException>(action);
+        }
+
+        [Test]
+        public void Constructor_NullColumns()
+        {
+            // Act
+            TestDelegate action = () => _ = new SaveAsMarkdownFileStreamWriter(
+                Stream.Null,
+                new SaveResultsAsMarkdownRequestParams(),
+                null
+            );
+
+            // Assert
+            Assert.Throws<ArgumentNullException>(action);
+        }
+
+        [Test]
+        public void Constructor_WithoutSelectionWithHeader_WritesHeaderWithAllColumns()
+        {
+            // Setup:
+            // ... Create a request params that has no selection made, headers should be printed
+            // ... Create a set of columns
+            // --- Create a memory location to store the output
+            var requestParams = new SaveResultsAsMarkdownRequestParams { IncludeHeaders = true };
+            var (columns, _) = GetTestValues(2);
+            byte[] output = new byte[8192];
+            using var outputStream = new MemoryStream(output);
+
+            // If: I construct a Markdown file writer
+            var writer = new SaveAsMarkdownFileStreamWriter(outputStream, requestParams, columns);
+            writer.Dispose();
+
+            // Then:
+            // ... It should have written a line
+            string[] lines = ParseWriterOutput(output, Environment.NewLine);
+            Assert.AreEqual(2, lines.Length);
```

Comment:
Tests should use [constraint style](https://docs.nunit.org/articles/nunit/writing-tests/assertions/assertion-models/constraint.html) (fluent) assertions. These are : 

- Easier to read
- Have better overloads (allow custom messages, necessary when there's multiple assertions in a single test to be able to tell what part exactly failed)
- Are harder to mess up the parameter order from (expected vs actual)